packages/flutter/lib/src/widgets/change_notifier_builder.dart

```diff
@@ -0,0 +1,145 @@
+  ///
+  /// The [child] is optional but is good practice to use if part of the widget
+  /// subtree does not depend on the value of the [changeNotifier].
+  ///
+  /// The [listenCondition] is also optional and it can be used to control when
+  /// the [builder] should be called.
+  const ChangeNotifierBuilder({
+    Key? key,
+    required this.changeNotifier,
+    required this.builder,
+    this.listenCondition,
+    this.child,
+  }) : super(key: key);
+
+  /// The [ChangeNotifier] whose value you depend on in order to build.
+  final T changeNotifier;
+
+  /// A [ChangeNotifierWidgetBuilder] which builds a widget depending on the
+  /// [changeNotifier]'s status.
+  ///
+  /// Can incorporate a [changeNotifier]-independent widget subtree
+  /// from the [child] parameter into the returned widget tree.
+  final ChangeNotifierWidgetBuilder<T> builder;
+
+  /// The listening condition for the [builder].
+  ///
+  /// This callback can be used to determine under which conditions the
+  /// [builder] function has to be invoked. For example, if this callback
+  /// returned false, the [builder] would never be called.
+  ///
+  /// By default, this is null so the [builder] il always called whenever the
+  /// [changeNotifier] notifies its listeners.
+  final bool Function()? listenCondition;
+
+  /// The child widget to pass to the [builder].
+  ///
+  /// If a [builder] callback's return value contains a subtree that does not
+  /// depend on [changeNotifier], it's more efficient to build that subtree once
+  /// instead of rebuilding it on every animation tick.
+  ///
+  /// If the pre-built subtree is passed as the [child] parameter, the
+  /// [ChangeNotifierBuilder] will pass it back to the [builder] function so
+  /// that it can be incorporated into the build.
+  ///
+  /// Using this pre-built child is entirely optional, but can improve
+  /// performance significantly in some cases and is therefore a good practice.
+  final Widget? child;
+
+  @override
+  State<ChangeNotifierBuilder<T>> createState() => _ChangeNotifierBuilderState<T>();
+}
+
+class _ChangeNotifierBuilderState<T extends ChangeNotifier> extends State<ChangeNotifierBuilder<T>> {
+  @override
+  void initState() {
+    super.initState();
+    widget.changeNotifier.addListener(_handleChange);
+  }
+
+  @override
+  void didUpdateWidget(covariant ChangeNotifierBuilder<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.changeNotifier != oldWidget.changeNotifier) {
+      oldWidget.changeNotifier.removeListener(_handleChange);
+      widget.changeNotifier.addListener(_handleChange);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.changeNotifier.removeListener(_handleChange);
+    super.dispose();
+  }
+
+  void _handleChange() {
+    if (widget.listenCondition != null) {
+      // Rebuilding only if the listening condition evaluates to true
+      if (widget.listenCondition!()) {
+        setState(() {});
```

Comment:
We shouldn't all empty `setState`s, but in this case I think we don't have choice, so we could at least document why we are doing this.

```suggestion
        setState(() {
          // The listenable's state is our build state, and it changed already.
        });
```